graphql-dxm-provider/src/main/java/org/jahia/modules/graphql/provider/dxm/instrumentation/JcrLiveOperationsTrackerInstrumentation.java

```diff
@@ -0,0 +1,103 @@
+import graphql.language.OperationDefinition;
+import graphql.schema.GraphQLArgument;
+import graphql.schema.GraphQLFieldDefinition;
+import graphql.schema.GraphQLNamedType;
+import org.jahia.modules.graphql.provider.dxm.node.GqlJcrMutation;
+import org.jahia.modules.graphql.provider.dxm.node.GqlJcrQuery;
+import org.jahia.modules.graphql.provider.dxm.node.NodeQueryExtensions.Workspace;
+import org.jahia.modules.graphql.provider.dxm.util.ContextUtil;
+import org.jahia.modules.graphql.provider.dxm.util.GqlTypeUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.servlet.http.HttpServletResponse;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+
+import static org.jahia.modules.graphql.provider.dxm.node.NodeQueryExtensions.WORKSPACE_PARAM_NAME;
+
+/**
+ * Custom instrumentation to track operations (queries and mutations) on the live JCR workspace.
+ */
+public class JcrLiveOperationsTrackerInstrumentation implements Instrumentation {
+    private static final Logger log = LoggerFactory.getLogger(JcrLiveOperationsTrackerInstrumentation.class);
+
+    private static final String LIVE_OPERATION_HEADER = "X-Jahia-Live-Operation";
+
+    @Override
+    public InstrumentationState createState(InstrumentationCreateStateParameters parameters) {
+        return new TrackerInstrumentationState();
+    }
+
+    @Override
+    public InstrumentationContext<ExecutionResult> beginField(InstrumentationFieldParameters parameters, InstrumentationState state) {
+        GraphQLFieldDefinition field = parameters.getField();
+        GraphQLNamedType unwrappedType = GqlTypeUtil.unwrapType(field.getType());
+        TrackerInstrumentationState customState = (TrackerInstrumentationState) state;
+        Workspace workspace = (Workspace) Optional.ofNullable(parameters.getExecutionStepInfo().getArgument(WORKSPACE_PARAM_NAME)).orElseGet(() -> getDefaultValue(parameters));
+        if (Workspace.LIVE.equals(workspace)) {
+            log.debug("'live' workspace used for field '{}'", field.getName());
+            switch (unwrappedType.getName()) {
+                case GqlJcrQuery.NAME:
+                    customState.liveOperation = OperationDefinition.Operation.QUERY;
+                    break;
+                case GqlJcrMutation.NAME:
+                    customState.liveOperation = OperationDefinition.Operation.MUTATION;
+                    break;
+                default:
+            }
+        }
+        return Instrumentation.super.beginField(parameters, state);
+    }
+
+    @Override
+    public CompletableFuture<ExecutionResult> instrumentExecutionResult(ExecutionResult executionResult, InstrumentationExecutionParameters parameters, InstrumentationState state) {
+        TrackerInstrumentationState jcrTrackerInstrumentationState = (TrackerInstrumentationState) state;
+        HttpServletResponse response = ContextUtil.getHttpServletResponse(parameters.getGraphQLContext());
```

Comment:
According to a comment found in: https://github.com/Jahia/graphql-core/blob/master/graphql-dxm-provider/src/main/java/org/jahia/modules/graphql/provider/dxm/instrumentation/JCRInstrumentation.java#L60-L63

It seem's it would be safe to check for httpServletResponse to be not null

Reply:
you're right, good catch!